From faa7fe94fba079b08cd515d66e2a277939883937 Mon Sep 17 00:00:00 2001
From: Franz-Josef Haider <frajo@frajo.fi>
Date: Sat, 20 Jun 2026 14:25:23 +0300
Subject: [PATCH] Don't fail due to -Werror=sizeof-array-div

---
 src/Android.mk | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/Android.mk b/src/Android.mk
index 1079838..1cd4d52 100644
--- a/src/Android.mk
+++ b/src/Android.mk
@@ -21,6 +21,7 @@ LOCAL_MODULE := libsndfile
 LOCAL_PROPRIETARY_MODULE := true
 LOCAL_MODULE_TAGS := optional
 LOCAL_PRELINK_MODULE := false
+LOCAL_CFLAGS += -Wno-error=sizeof-array-div
 
 include $(BUILD_SHARED_LIBRARY)
 
-- 
2.55.0

